src/de/mod.rs

```diff
@@ -370,12 +416,69 @@ where
     }
 
     fn next(&mut self) -> Result<DeEvent<'de>, DeError> {
+        // Replay skipped or peeked events
+        #[cfg(feature = "overlapped-lists")]
+        if let Some(event) = self.read.pop_front() {
+            return Ok(event);
+        }
+        #[cfg(not(feature = "overlapped-lists"))]
         if let Some(e) = self.peek.take() {
             return Ok(e);
         }
         self.reader.next()
     }
 
+    /// Extracts XML tree of events from and stores them in the skipped events
+    /// buffer from which they can be retrieved later. You MUST call
+    /// [`Self::start_replay()`] after calling this to give access to the skipped
+    /// events and release internal buffers.
+    #[cfg(feature = "overlapped-lists")]
+    fn skip(&mut self) -> Result<(), DeError> {
+        let event = self.next()?;
+        self.write.push_back(event);
+        match self.write.back() {
+            // Skip all subtree, if we skip a start event
+            Some(DeEvent::Start(e)) => {
+                let end = e.name().to_owned();
+                let mut depth = 0;
+                loop {
+                    let event = self.next()?;
+                    match event {
+                        DeEvent::Start(ref e) if e.name() == end => {
+                            self.write.push_back(event);
+                            depth += 1;
+                        }
+                        DeEvent::End(ref e) if e.name() == end => {
+                            self.write.push_back(event);
+                            if depth == 0 {
+                                return Ok(());
+                            }
+                            depth -= 1;
+                        }
+                        _ => self.write.push_back(event),
+                    }
+                }
+            }
+            _ => Ok(()),
+        }
+    }
+
+    /// Moves all buffered events to the end of [`Self::write`] buffer and swaps
+    /// read and write buffers.
+    ///
+    /// After calling this method, [`Self::peek()`] and [`Self::next()`] starts
+    /// return events that was skipped previously by calling [`Self::skip()`],
+    /// and only when all that events will be consumed, the deserializer starts
+    /// to drain events from underlying reader.
+    ///
+    /// This method MUST be called if any number of [`Self::skip()`] was called
+    /// after [`Self::new()`] or `start_replay()` or you'll lost events.
+    #[cfg(feature = "overlapped-lists")]
```

Comment:
Just curious about the use case, is this basically for situations where you're parsing messy handwritten XML / HTML?  Generally I would think machine-generated XML wouldn't have this problem.